feat: keys include host-port-suffix

matt2e opened the pull request:
Fixes #1047
Fixes #1045

Changes `ControllerKey` and `RunnerKey` to have the format:
- `<r/c>-<hostname>-<port>-<4bytehex>` for passing around
- `<hostname>-<port>-<4bytehex>` in the db
- non production environments use shorter keys: `r-<incrementing-digits>` and `<incrementing digits>` in db

matt2e commented on backend/controller/sql/types.go:
```diff
@@ -1,61 +1,10 @@
-	"github.com/oklog/ulid/v2"
 )
 
-type NullKey = optional.Option[Key]
```
We have no more uuid/Key columns so this code was all being unused. Is it ok to remove it like I have or are we wanting to keep it for other UUID stuff coming down the road?

alecthomas replied:
Nope nuke it. We can always restore it from history if we need it.

matt2e commented on backend/controller/sql/schema/001_init.sql:
```diff
@@ -101,7 +101,7 @@ CREATE TABLE runners
 (
     id                  BIGINT       NOT NULL GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
     -- Unique identifier for this runner, generated at startup.
-    key                 UUID UNIQUE  NOT NULL,
```
Are we worried about changing schema without any migration?

matt2e replied:
Not at this point

matt2e commented on backend/protos/xyz/block/ftl/v1/ftl.proto:
```diff
@@ -146,7 +146,6 @@ enum RunnerState {
 }
 
 message RegisterRunnerRequest {
-  // UUID representing the runner instance.
```
I think this is supposed to update the identifical comment in ftl_pb.ts and ftl.pb.go? Running `bit` didn't seem to trigger anything

matt2e commented on backend/controller/scaling/localscaling/local_scaling.go:
```diff
@@ -40,6 +42,7 @@ func NewLocalScaling(portAllocator *bind.BindAllocator, controllerAddresses []*u
 		runners:             map[model.RunnerKey]context.CancelFunc{},
 		portAllocator:       portAllocator,
 		controllerAddresses: controllerAddresses,
+		prevRunnerSuffix:    0, // first runner will have a key of r-0001
```
Use optional type

matt2e replied:
Ok to start at 1

alecthomas commented on internal/model/keys.go:
```diff
@@ -5,82 +5,145 @@ import (
+	}
+}
+func ParseRunnerKey(key string) (RunnerKey, error)   { return parseKey[RunnerKey](key, true) }
+func ParseRunnerDBKey(key string) (RunnerKey, error) { return parseKey[RunnerKey](key, false) }
```
You shouldn't need this function once the DB types are added.

alecthomas commented on internal/model/keys.go:
```diff
@@ -5,82 +5,145 @@ import (
 func (d keyType[T]) MarshalText() ([]byte, error) { return []byte(d.String()), nil }
 func (d *keyType[T]) UnmarshalText(bytes []byte) error {
-	id, err := parseKey[keyType[T]](string(bytes))
+	fmt.Printf("marshal text: %s\n", string(bytes))
```
Remove

alecthomas commented on internal/model/keys_test.go:
```diff
@@ -8,9 +8,49 @@ import (
+		{key: NewRunnerKey("r-hostwithprefixandfakeport-80", "80"), strPrefix: "r-r-hostwithprefixandfakeport-80-80-", valuePrefix: "r-hostwithprefixandfakeport-80-80-"},
+
+		// Local Keys
+		{key: NewLocalRunnerKey(0), str: "r-0000", value: "0000"},
```
Nice

alecthomas commented on internal/model/keys.go:
```diff
@@ -2,85 +2,157 @@
+	if d.Hostname == "" {
+		return fmt.Sprintf("%s%04d", prefix, d.Suffix)
+	}
+	return fmt.Sprintf("%s%s-%s-%04d", prefix, d.Hostname, d.Port, d.Suffix)
```
Let's hex encode the suffix with `%08x` so it's consistent with DeploymentName.

alecthomas commented on internal/model/keys.go:
```diff
@@ -2,85 +2,157 @@
-func NewControllerKey() ControllerKey                      { return ControllerKey(ulid.Make()) }
-func ParseControllerKey(key string) (ControllerKey, error) { return parseKey[ControllerKey](key) }
+func NewControllerKey(hostname string, port string) ControllerKey {
+	suffix, err := rand.Int(rand.Reader, big.NewInt(10000))
```
Let's use 4 bytes (4294967296) of randomness so it's more consistent with DeploymentName.

alecthomas commented on internal/model/keys.go:
```diff
@@ -2,85 +2,157 @@
+	case len(components) == 1:
+		//style: [<kind>-]<suffix>
+
+		suffix, err := strconv.Atoi(components[len(components)-1])
```
When you switch to hex encoding you can use `strconv.ParseInt()` with a base of 16 here.

alecthomas reviewed (changes requested):
LGTM after changing the suffix to more bits of hex-encoded randomness

matt2e commented:
@alecthomas i switched to 4 bytes of randomness in production keys, moved Suffix to being a string so it can hold hex or incremental ints, and now we can have 0 based indexes again.
